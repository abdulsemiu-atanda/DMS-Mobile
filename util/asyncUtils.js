import sendRequest from '../requests/requestUtil'

/**
 * Constructs async actions constants
 * @param {string} baseName
 * @returns {Object}
 */
export const asyncActionNames = baseName => ({
  failure: `${baseName}_FAILURE`,
  loading: `${baseName}_LOADING`,
  success: `${baseName}_SUCCESS`
})

/**
 * Constructs Redux async actions
 * @param {string} actionName
 * @returns {Object}
 */
export const asyncActions = (actionName) => ({
  loading: bool => ({
    type: asyncActionNames(actionName).loading,
    data: bool
  }),
  failure: (bool, error) => ({
    type: asyncActionNames(actionName).failure,
    data: {error, status: bool}
  }),
  success: data => ({
    type: asyncActionNames(actionName).success,
    data
  })
})

/**
 * Dispatches Redux Actions
 * @async asyncRequest
 * @param {string} ACTION_NAME - The name of action to be dispatched
 * @param {string} endpoint - The endpoint to resource
 * @param {string} method - HTTP method
 * @param {Object} data - HTTP request body
 * @param {string} token - Authentication token for request
 * @returns {Function} - Function that dispatches Redux Actions
 */

export const asyncRequest = (ACTION_NAME, endpoint, method, data, token) => dispatch => {
  dispatch(asyncActions(ACTION_NAME).loading(true))
  sendRequest(endpoint, method, data, token)
    .then(res => res.json())
    .then(res => {
      dispatch(asyncActions(ACTION_NAME).success(res))
      dispatch(asyncActions(ACTION_NAME).loading(false))
    })
    .catch(err => dispatch(asyncActions(ACTION_NAME).failure(true, err)))
}
