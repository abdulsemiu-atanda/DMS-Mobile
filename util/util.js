import {
  emailCorrect,
  emailError,
  fullnameCorrect,
  fullnameError,
  passwordMatch,
  passwordMatchError,
  usernameCorrect,
  usernameError
} from '../actions/formActions'

export const validateFields = data => {
  return dispatch => {
    Object.keys(data).forEach(field => {
      switch (field) {
        case 'email':
          return /^\S+@\S+$/g.test(data[field]) ? dispatch(emailCorrect(true)) : dispatch(emailError(true))
        case 'username':
          return data[field].trim() === '' ? dispatch(usernameError(true)) : dispatch(usernameCorrect(true))
        case 'password':
          return data.password === data.confirmPassword ? dispatch(passwordMatch(true)) : dispatch(passwordMatchError(true))
        case 'fullname':
          return data[field].trim() === '' ? dispatch(fullnameError(true)) : dispatch(fullnameCorrect(true))
        default:
          break
      }
    })
  }
}