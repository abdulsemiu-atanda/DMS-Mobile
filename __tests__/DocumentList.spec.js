import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import {fromJS} from 'immutable'
import moment from 'moment'

import DocumentList from '../components/shared/DocumentList.react'
import {noop} from '../util/util'

const date = new Date(1330688329321)

const publicDocuments = [
  {
    title: 'Scooby Doo',
    access: 'public',
    content: 'Scooby Doo is a beloved brown dog.',
    updatedAt: date.toISOString(),
    createdAt: date.toISOString()
  },
  {
    title: 'Sharpie',
    access: 'public',
    content: 'Sharpie is one of the perfectly crafted products.',
    updatedAt: date.toISOString(),
    createdAt: date.toISOString()
  }
]
const props = {
  documents: fromJS(publicDocuments),
  navigation: {
    navigate: noop,
    state: {}
  },
  screen: 'All'
}

describe('DocumentList', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<DocumentList {...props} />).toJSON()
    const component = tree.children[0].children[0].children[0].children[0].children[0].children

    expect(tree).toMatchSnapshot()
    expect(component[0].type).toEqual('Text')
    expect(component[0].children[0]).toEqual(publicDocuments[0].title)
    expect(component[1].children[0]).toEqual(publicDocuments[0].content)
    expect(component[3].children[1])
      .toEqual(moment(publicDocuments[0].updatedAt).format('MMM Do YY'))
    expect(component[3].children[3])
      .toEqual(moment(publicDocuments[0].createdAt).format('MMM Do YY'))
  })

  it('renders correctly when user is viewing their documents', () => {
    const userViewProps = {
      ...props,
      screen: 'Collection'
    }
    const tree = renderer.create(<DocumentList {...userViewProps} />).toJSON()
    const categoryView = tree.children[1].children[0].children[0].children[0].children[0]

    expect(tree).toMatchSnapshot()
    expect(tree.children[0].children[0].children[0]).toEqual('ACCESS')
    expect(categoryView.children[0]).toEqual('Public')
    expect(parseInt(categoryView.children[2])).toEqual(userViewProps.documents.size)
  })

  it('renders correctly when there are no documents', () => {
    const noDocProps = {
      ...props,
      documents: fromJS([])
    }
    const tree = renderer.create(<DocumentList {...noDocProps} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.children[1].children[0])
      .toMatch('Oops! there are no public documents currently available for viewing.')
  })
})
