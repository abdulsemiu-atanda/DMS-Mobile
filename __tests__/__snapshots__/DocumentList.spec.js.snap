// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentList renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 7,
    }
  }
>
  <RCTScrollView
    dataSource={
      ListViewDataSource {
        "items": 2,
      }
    }
    enableEmptySections={true}
    renderRow={[Function]}
    renderScrollComponent={[Function]}
    style={
      Object {
        "backgroundColor": "#f7f7f7",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "borderRadius": 5,
            "marginBottom": 5,
            "width": 750,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "height": 266.8,
                "padding": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 15,
                  "fontWeight": "700",
                  "marginBottom": 2,
                }
              }
            >
              Scooby Doo
            </Text>
            <Text
              numberOfLines={3}
              style={
                Object {
                  "marginBottom": 106.72,
                }
              }
            >
              Scooby Doo is a beloved brown dog.
            </Text>
            <View
              style={
                Object {
                  "borderBottomColor": "#bbb",
                  "borderBottomWidth": 1,
                  "marginBottom": 2,
                }
              }
            />
            <Text
              style={
                Object {
                  "color": "#bbb",
                  "fontSize": 10,
                }
              }
            >
              Modified: 
              Mar 2nd 12
               / Created: 
              Mar 2nd 12
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "borderRadius": 5,
            "marginBottom": 5,
            "width": 750,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "height": 266.8,
                "padding": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 15,
                  "fontWeight": "700",
                  "marginBottom": 2,
                }
              }
            >
              Sharpie
            </Text>
            <Text
              numberOfLines={3}
              style={
                Object {
                  "marginBottom": 106.72,
                }
              }
            >
              Sharpie is one of the perfectly crafted products.
            </Text>
            <View
              style={
                Object {
                  "borderBottomColor": "#bbb",
                  "borderBottomWidth": 1,
                  "marginBottom": 2,
                }
              }
            />
            <Text
              style={
                Object {
                  "color": "#bbb",
                  "fontSize": 10,
                }
              }
            >
              Modified: 
              Mar 2nd 12
               / Created: 
              Mar 2nd 12
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`DocumentList renders correctly when there are no documents 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "grey",
          "fontSize": 70,
        },
        undefined,
        Object {
          "fontFamily": "Ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      ]
    }
  >
    ï†«
  </Text>
  <Text
    style={
      Object {
        "fontFamily": "Verdana",
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Oops! there are no public documents currently available for viewing.        Please check again later or start by adding a public document.
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignContent": "center",
        "alignSelf": "center",
        "backgroundColor": "#0983ff",
        "padding": 7,
        "width": 300,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#aacfff",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      ADD DOCUMENT
    </Text>
  </View>
</View>
`;

exports[`DocumentList renders correctly when user is viewing their documents 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 7,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 10,
        "width": 750,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Verdana",
          "fontSize": 12,
        }
      }
    >
      ACCESS
    </Text>
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "#bbb",
            "borderBottomWidth": 1,
            "marginBottom": 2,
          },
          Object {
            "marginBottom": 8,
            "width": 600,
          },
        ]
      }
    />
  </View>
  <RCTScrollView
    dataSource={
      ListViewDataSource {
        "items": 2,
      }
    }
    enableEmptySections={true}
    renderRow={[Function]}
    renderScrollComponent={[Function]}
    style={
      Object {
        "backgroundColor": "#f7f7f7",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "column",
            "height": 106.72,
            "justifyContent": "space-between",
            "padding": 10,
            "width": 750,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "Verdana",
                "fontSize": 12,
              }
            }
          >
            Public
             (
            2
            )
          </Text>
        </View>
        <View
          style={
            Object {
              "borderBottomColor": "#bbb",
              "borderBottomWidth": 1,
              "marginBottom": 2,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
